### PR TITLE
Reservations plugin

### PR DESCRIPTION
**Pull request by shawnchin**
This is a port of [Jicofo's reservation system](https://github.com/jitsi/jicofo/blob/master/doc/reservation.md) to a prosody module following conversation with @damencho in this thread: https://community.jitsi.org/t/gatekeeping-room-creation-from-prosody-module/89720

The flow turned out slightly more complex than I anticipated, so hopefully the following diagram helps clarify the design:
mod_reservations

### CODE REVIEW

**Comment by shawnchin** [comment]
@damencho Just pushed changes to support HTTP retries. Apart from cleaner user experience when room expires, this should covert he rest of the TODOs.

**Comment by shawnchin** [comment]
> Why do we need to enable/disable it? Can that always be on?

I'm happy for it to always be on, but as far as I can tell, that's not how Jicofo's reservation system behaves and I thought it work keeping it compatible.

We could leave it enabled by default and leave the option in for people to disable if it causes trouble with their existing setup?

**Comment by shawnchin** [comment]
Having reviewed again what I've done, I totally see your point. This would be so much cleaner and less prone to mistakes (spotted a couple) if that was always on since we could just rewrite the room jid right from the beginning.

Making that change now.

**Comment by TheDOC123** [comment]
Two findings:

1. Line 508: module.log instead of module:log.
2. The DELETE HTTP request is not done when the room expires only if the room is destroyed in other ways?

I'm having trouble getting this up and running the way it's supposed to be. I will do some more testing in the next days and get back with a report. But thanks for the work up until this point.

File: resources/prosody-plugins/mod_reservations.lua [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+    local room;
+    for _, room_jid in ipairs(expired) do
+        module:log("debug", "Dropped expired reservation for %s", room_jid);
+        res = reservations[room_jid]
+        reservations[room_jid] = nil;
+
+        room = get_room_from_jid(room_jid);
+        if room then
+            -- Close room if still active (reservation duration exceeded)
+            module.log("info", "Room exceeded reservation duration. Terminating %s", room_jid);
```
Comment (TheDOC123) [inline resources/prosody-plugins/mod_reservations.lua:508]: module.log instead of module:log

**Comment by shawnchin** [comment]
> 2. The DELETE HTTP request is not done when the room expires only if the room is destroyed in other ways?

Oh, good spot! 

For some reason I was relying on `muc-room-destroyed` event handler to call DELETE in this case too, but of course that's not going to happen since I already dropped the reservation metadata in the expiry handler. 

I'll work on a fix.

File: resources/prosody-plugins/mod_reservations.lua (hunk 1 above)
Comment (shawnchin) [reply resources/prosody-plugins/mod_reservations.lua]: Fixed in latest push, along with issue with DELETE not called on room expiry.

**Comment by damencho** [comment]
This is not forgotten, but I'm so sorry, I just realized its been 4 months already :( I will try to squeeze it in the following weeks ...

File: resources/prosody-plugins/mod_reservations.lua [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+    end
+
+    local room;
+    for _, room_jid in ipairs(expired) do
+        room = get_room_from_jid(room_jid);
+        if room then
+            -- Close room if still active (reservation duration exceeded)
+            module:log("info", "Room exceeded reservation duration. Terminating %s", room_jid);
+            room:clear();
+            room:destroy();
```
Comment (damencho) [inline resources/prosody-plugins/mod_reservations.lua:504]: Hey, @shawnchin do you remember why you added room:clear here?

Dropping the clear and just changing `room:destroy(nil, "Scheduled conference duration exceeded.");` fixes the destroy problem.

**Comment by damencho** [comment]
I just added a small fix for the room destroy and will merge it. Thank you @shawnchin for your contribution and sorry for the delay.

@shawnchin one, more question, do you allow me to use the image you added here in the PR when moving the documentation from jicofo to the handbook?

**Review by damencho — APPROVED** [review]

**Comment by damencho** [comment]
@shawnchin and if you are ok with the image, by any chance do you have some source of the drawing :)

**Comment by damencho** [comment]
Awsome job :)

**Comment by shawnchin** [comment]
> @shawnchin and if you are ok with the image, by any chance do you have some source of the drawing :)

Please do feel free to use the image. 

As for the source, I used internally hosted lucidchart for the drawing; I will try later to see if it can be exported to an editable format.

**Comment by shawnchin** [comment]
> Awsome job :)

Thanks! And thanks for the review + merge 🎉

**Comment by shawnchin** [comment]
Here's a zip with SVG and Visio exports of the image. Not sure how well the formatting is preserved or how editable it is going to be.

[mod_reservations_images.zip](https://github.com/jitsi/jitsi-meet/files/6477229/mod_reservations_images.zip)

Also happy to make tweaks to the image for you if you need some changes down the road.

File: resources/prosody-plugins/mod_reservations.lua (hunk 2 above)
Comment (shawnchin) [reply resources/prosody-plugins/mod_reservations.lua]: Don't recall any good reason. It's likely I picked it up from somewhere in the community forum.

Thanks for fixing :)

**Comment by damencho** [comment]
Thank you. https://github.com/jitsi/handbook/pull/193

**Comment by L3o-pold** [comment]
Am I missing something or it's needed to add undocumented `focus_component` or `muc_mapper_domain_base` option (https://github.com/jitsi/jitsi-meet/pull/8386/files#diff-c76dcaac2995d3881f36fd99a4e228c25a4155740e08bcdb365d5d01d62fe0f1R99) ?

Could be nice to add a note on the handbook for this and/or add it here https://github.com/jitsi/docker-jitsi-meet/blob/master/prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua

**Comment by damencho** [comment]
It is already added in the default configs https://github.com/jitsi/jitsi-meet/blob/6ca61e03fdbda5b5474d0db19872bc7f7a4bc7c9/doc/debian/jitsi-meet-prosody/prosody.cfg.lua-jvb.example#L4
focus_component is not required as it has a default value.
If something is missing in docker repo fill free to create a PR.

**Comment by L3o-pold** [comment]
Some change from Jicofo reservation system (that may break for you during the migration)

- mail_owner is required in the API return
- duration is no more sent to the API

**Comment by L3o-pold** [comment]
A major issue is that when using an API url `reservations_api_prefix` with a service name in docker or kubernetes like this `http://my-awsome-api` you will have an error code 0 (in your callback) with the message `unable to resolve service`.

**Comment by shawnchin** [comment]
> Some change from Jicofo reservation system (that may break for you during the migration)
> 
> * mail_owner is required in the API return
> * duration is no more sent to the API

Did Jicofo send duration to the API? I thought it was the other way round where the API returns the duration to Jicofo as per the docs: https://github.com/jitsi/jicofo/blob/be025fbde8e077784f21fe0fcf5b041716961b56/doc/reservation.md#conference-allocation 

I'm afraid the prosody module was build using the documentation as specs, so any deviation from the docs could indeed break during migration.

**Comment by shawnchin** [comment]
> A major issue is that when using an API url `reservations_api_prefix` with a service name in docker or kubernetes like this `http://my-awsome-api` you will have an error code 0 (in your callback) with the message `unable to resolve service`.

Pure guesswork on my end, but this sounds more like a DNS resolution issue. Quoting [docs](https://docs.docker.com/config/containers/container-networking/#dns-services) (and not backed by any real experience on my part), could this be down to the use of bridge network so embedded DNS server is not used and custom service name cannot be resolved?

**Comment by L3o-pold** [comment]
> > Some change from Jicofo reservation system (that may break for you during the migration)
> > 
> > * mail_owner is required in the API return
> > * duration is no more sent to the API
> 
> Did Jicofo send duration to the API? I thought it was the other way round where the API returns the duration to Jicofo as per the docs: https://github.com/jitsi/jicofo/blob/be025fbde8e077784f21fe0fcf5b041716961b56/doc/reservation.md#conference-allocation
> 
> I'm afraid the prosody module was build using the documentation as specs, so any deviation from the docs could indeed break during migration.

It's not a big deal, but it may be important to some user to be warn before the migration.



> > A major issue is that when using an API url `reservations_api_prefix` with a service name in docker or kubernetes like this `http://my-awsome-api` you will have an error code 0 (in your callback) with the message `unable to resolve service`.
> 
> Pure guesswork on my end, but this sounds more like a DNS resolution issue. Quoting [docs](https://docs.docker.com/config/containers/container-networking/#dns-services) (and not backed by any real experience on my part), could this be down to the use of bridge network so embedded DNS server is not used and custom service name cannot be resolved?

No because a curl works on the same service name `http://my-awsome-api`. Anyway my workaround is to use a real FQDN but I prefer to warn it here if someone see the same issue and can save some debug time.

**Comment by jbg** [comment]
This doesn't seem to work if you configure focus directly as a JID in the frontend's `config.js`, like `focus: "focus@auth.example.org/focus"`. If you go via the client proxy with `focus: "focus.example.org"` it works fine.

Not a major issue but just mentioning it in case anyone else hits it. If there is a plan to eventually remove the need for the client proxy then maybe this should be extended to also work in this case.

**Comment by shawnchin** [comment]
Indeed it defaults to intercepting IQs addressed to "focus.MUC_DOMAIN" if not overridden. 

Does it by any chance work if you set `focus_component = "focus@auth.example.org/focus"` in the domain config? I've never tested that use case and simply guessing based on:

https://github.com/jitsi/jitsi-meet/blob/cff4ed83f695a7d9d7a1bd1a99bde0319265ee72/resources/prosody-plugins/mod_reservations.lua#L99-L107

**Comment by pascuflow** [comment]
Possible minor bug, not sure the ` if res then  -- just in case event triggered more than once?` line 560 is needed, I have a 10 sec expiry_check_period and the DELETE call never happens because the evict_expired_reservations runs faster and sets reservations[room_jid] = nil.

**Comment by shawnchin** [comment]
But "reservations[room_jid] = nil" is only called if room no longer exists -- e.g. when all occupants left before expiry -- so I'm afraid I can't yet see how a short `expiry_check_period` could result in DELETE not being called. 

https://github.com/jitsi/jitsi-meet/blob/60dcac96a60583d3776cd0ec0697674e514ab23d/resources/prosody-plugins/mod_reservations.lua#L496-L510

So when `evict_expired_reservations` runs, rooms that have expired is either still running and gets  destroyed (which triggers `room_destroyed` function that will send the DELETE call), or it no longer exists (everyone left, and `room_destroyed` would have been triggered then) so it is dropped from the reservations table.

The scenario in which this could break would be if room is destroyed and `evict_expired_reservations` is somehow dispatched before `muc-room-destroyed` hook is triggered. No idea if this is a plausible scenario on prosody.

**Comment by pascuflow** [comment]
Well having a short expiry_check_period would increase the chance of what you mention in the last paragraph, it calls evict_expired_reservations more frequently. In any case, what I mentioned fixed the bug, everything is working fine now.

**Comment by shawnchin** [comment]
Glad it's sorted on your end.

Don't think `conflict_id = room.jid` will work in general though since according to the specs conflict_id is some value returned by the API call and not tied to room jid.

**Comment by pascuflow** [comment]
Yup, but I found it is a substring, so I just slice it on my end.

**Pull request merged**